### DISCUSSION TITLE
STL reader does not give access to `STLSolidLabeling`

### DISCUSSION THREAD

**Discussion by gnikit** [pyvista/pyvista#5042] [Q&A] [answered]
With the deprecation of attributes to readers, what is the recommended way of reading in `STLSolidLabeling` from an STL?
Previously one would do:

```py
import pyvista as pv
mesh = pv.read("labelled_solid.stl", attrs={"SetScalarTags": True})
print(f"Array names: {mesh.array_names}")
```
```
Array names: ['STLSolidLabeling']
```

Using a reader does not give access to the necessary information:

```py
import pyvista as pv
reader: pv.STLReader = pv.get_reader("labelled_solid.stl")
mesh: pv.PolyData = reader.read()
print(f"Array names: {mesh.array_names}")
```

Is this functionality intentionally being deprecated?

**Comment by MatthewFlamm** [comment, marked as answer]
Since I don't have your data I cannot test this, but try:

```py
import pyvista as pv
reader: pv.STLReader = pv.get_reader("labelled_solid.stl")
reader.reader.SetScalarTags(True)
mesh: pv.PolyData = reader.read()
print(f"Array names: {mesh.array_names}")
```

Ideally someone can write a PR for exposing this vtk method into the pyvista class so it would be something like:

```py
import pyvista as pv
reader: pv.STLReader = pv.get_reader("labelled_solid.stl")
reader.set_scalar_tags(True)
mesh: pv.PolyData = reader.read()
print(f"Array names: {mesh.array_names}")
```

**Reply by gnikit** [reply]
Thanks @MatthewFlamm that worked! Ideally shouldn't the method be exposed via the constructor too?

**Reply by MatthewFlamm** [reply]
My opinion: The current state makes the Reader classes more vtk like.  There are some options where you may need to inspect the header data before setting a value, which is not possible in the constructor. For maintenance of the code, it is also much easier as the constructor code is the same for nearly all Readers.

Having said that, we have set the defaults in PyVista to be different than vtk when it reads in more data.  For example reading in both cell and point data by default.  If this is the case here, setting this flag to True reads in scalar data, Id be +1 for this to be set as the default behavior.

**Reply by gnikit** [reply]
Thanks for the detailed response, I will see if I can get a PR underway in the next couple of weeks.

**Reply by MatthewFlamm** [reply]
Please let me know if you have any questions, or open up a draft PR is even better.  Ideally we like to have tests for additions/changes to the code.  The data we use for testing reading files is located at the pyvista/vtk-data repository.  The current test for `STLReader` is here.  If there is another data file in `vtk-data` repository that is needed with the right kind of data, it can be some work to add it, but we can help.

https://github.com/pyvista/pyvista/blob/e574da02bb2c2d7aafb2e8e15f55024505087b56/tests/core/test_reader.py#L298-L305
